File: Recursion/src/main/python/solution77.py
from typing import List


class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        self.n = n
        self.k = k
        self.sub_ans = []
        self.ans = []
        self.recur(1)
        return self.ans

    def recur(self, i):
        # if i == n:
        if i == self.n + 1:
            if len(self.sub_ans) == self.k:
                self.ans.append(self.sub_ans[:])
            return

        self.recur(i + 1)

        self.sub_ans.append(i)

        self.recur(i + 1)

        self.sub_ans.pop()


s = Solution()
res = s.combine(4, 2)
print(res)

File: Recursion/src/main/java/Solution77.java

import java.util.*;

public class Solution77 {
    public List<List<Integer>> combine(int n, int k) {
        this.k = k;
        this.n = n;
        recur(1);
        return res;
    }

    int k;
    int n;
    List<List<Integer>> res = new ArrayList<>();
    List<Integer> subRes = new ArrayList<>();

    public void recur(int i) {
        if (i == n + 1) {
            if (subRes.size() == k) {
                res.add(new ArrayList<>(subRes));
            }
            return;
        }

        recur(i + 1);

        subRes.add(i);

        recur(i + 1);

        subRes.remove(subRes.size() - 1);

    }

    public static void main(String[] args) {
        Solution77 s = new Solution77();
        int n = 4;
        int k = 2;
        List<List<Integer>> res = s.combine(n, k);
        System.out.println(res);
    }
}
